File: java8/src/main/java/com/shekhargulati/ninetynine_problems/_01_lists/P19.java
package com.shekhargulati.ninetynine_problems._01_lists;

import java.util.List;
import java.util.Map;

/**
 * (**) Rotate a list N places to the left
 */
public class P19 {

    public static <T> List<T> rotate(List<T> list, int n) {
        n = n < 0 ? list.size() + n : n;
        Map<Boolean, List<T>> split = P17.split(list, n);
        List<T> result = split.get(false);
        result.addAll(split.get(true));
        return result;
    }
}

File: scala/src/main/scala/com/shekhargulati/ninetynine_problems/_01_lists/P19.scala
package com.shekhargulati.ninetynine_problems._01_lists

import scala.annotation.tailrec

/**
  * (**) Rotate a list N places to the left
  */
object P19 {

  /*
  Using split function defined in P17
   */
  def rotate[T](list: List[T], n: Int): List[T] = {
    val (first, second) = P17.split(list, makeNPositive(list, n))
    second ++: first
  }

  /*
    Using recursion
     */
  def rotateR[T](list: List[T], n: Int): List[T] = {
    @tailrec
    def rotate(list: List[T], result: List[T], cur: Int, n: Int): List[T] = list match {
      case x :: xs if cur < n => rotate(xs, result :+ x, cur + 1, n)
      case l => l ++ result
    }
    rotate(list, List(), 0, makeNPositive(list, n))
  }

  private def makeNPositive[T](list: List[T], n: Int): Int = if (n < 0) list.length + n else n


}
